File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [1] * n
        i, j = 1, 0
        while i < n:
            j = i-1
            while j >= 0:
                if j >=0 and nums[j] < nums[i]:
                    dp[i] = max(dp[i], 1 + dp[j])
                j-=1
            i += 1
        return max(dp)

File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.go
func lengthOfLIS(nums []int) int {
    if len(nums) == 0 { return 0 }

    dp := make([]int, len(nums))

    res := 1
    for i := len(nums)-1; i >= 0; i-- {
        dp[i] = 1
        for j := i + 1; j < len(nums); j++ {
            if nums[j] > nums[i] {
                dp[i] = max(dp[i], 1 + dp[j])
            }
        }
        res = max(res, dp[i])
    }
    return res
}

func max(x, y int) int {
    if x < y { return y }
    return x
}